pkg/querier/queryrange/ordering.go

```diff
@@ -44,11 +44,17 @@ func (a byDir) EntriesCount() (n int) {
 
 func (a byDir) merge() []logproto.Entry {
 	result := make([]logproto.Entry, 0, a.EntriesCount())
-
-	sort.Sort(a)
-	for _, m := range a.markers {
-		result = append(result, m...)
+	
+	for _, m := range a.markers { 
+		result = append(result, m...) 
 	}
+
+	sort.SliceStable(result, func(i, j int) bool {
+		if a.direction == logproto.BACKWARD {
+			return result[i].Timestamp.After(result[j].Timestamp)
+		}
+		return result[i].Timestamp.Before(result[j].Timestamp)
+	})
```

Comment:
Entries are already sorted at this point in time. Either by timestamp ascending in case of FORWARD direction, or by timestamp descending in case of BACKWARD direction.

The `merge()` is only called in the `mergeOrderedNonOverlappingStreams()` function https://github.com/grafana/loki/blob/main/pkg/querier/queryrange/codec.go#L1617 which assumes ordered entries.

I would not know a condition under which entries are not ordered by timestamp (ascending or descending).

Can you verify that the entries are not ordered by printing their timestamp in a custom build?